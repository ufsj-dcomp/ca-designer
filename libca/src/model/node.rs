#![allow(non_local_definitions)] // Hash32 is outdated, need to fix fchashmap too

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Eq, Hash, Debug, Default)]
pub struct NodeId(pub(crate) usize);

impl NodeId {
    #[inline]
    pub fn from_index(idx: usize) -> Self {
        Self(idx)
    }

    #[inline]
    pub fn as_index(self) -> usize {
        self.0
    }
}

#[derive(Serialize, Deserialize)]
pub struct Node(pub(crate) String);
