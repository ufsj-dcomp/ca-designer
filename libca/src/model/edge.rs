use serde::{Deserialize, Serialize};

use crate::state_map::StateMap;

use super::node::NodeId;

#[derive(Serialize, Deserialize, Clone, Copy)]
pub struct EdgeId(pub(crate) u32);

#[derive(Serialize, Deserialize)]
pub struct Edge {
    pub(crate) id: EdgeId,
    pub(crate) name: String,
    pub(crate) from_node: NodeId,
    pub(crate) to_node: NodeId,
    /// Conditions must ALL match, so there's an implicit `&&`
    /// operator between any pair of conditions.
    pub(crate) conditions: Vec<Condition>,
}

impl Edge {
    pub fn transition(&self, node_id: NodeId, neighbors: &StateMap) -> Option<NodeId> {
        (self.from_node == node_id
            && self
                .conditions
                .iter()
                .all(|cond| cond.is_satisfied(neighbors)))
        .then_some(self.to_node)
    }
}

#[derive(Serialize, Deserialize)]
pub struct Condition {
    pub(crate) left: Value,
    pub(crate) operand: Operand,
    pub(crate) right: Value,
}

impl Condition {
    fn is_satisfied(&self, neighbors: &StateMap) -> bool {
        let left = self.left.to_absolute(neighbors);
        let right = self.right.to_absolute(neighbors);
        self.operand.compare(left, right)
    }
}

#[derive(Serialize, Deserialize, Clone, Copy)]
// #[serde(tag = "type", content = "val")]
pub enum Value {
    Absolute(u32),
    PopulationCount(NodeId),
}

impl Value {
    #[inline]
    fn to_absolute(self, neighbors: &StateMap) -> u32 {
        match self {
            Value::Absolute(abs) => abs,
            Value::PopulationCount(node_id) => neighbors.get_count(node_id) as u32,
        }
    }
}

#[derive(Serialize, Deserialize, Clone, Copy)]
pub enum Operand {
    /// `==`
    Equal,
    /// `>`
    Greater,
    /// `>=`
    GreaterOrEqual,
    /// `<`
    Less,
    /// `<=`
    LessOrEqual,
    /// `!=`
    Different,
}

impl Operand {
    #[inline]
    pub fn compare(self, left: u32, right: u32) -> bool {
        match self {
            Operand::Equal => left == right,
            Operand::Greater => left > right,
            Operand::GreaterOrEqual => left >= right,
            Operand::Less => left < right,
            Operand::LessOrEqual => left <= right,
            Operand::Different => left != right,
        }
    }
}
