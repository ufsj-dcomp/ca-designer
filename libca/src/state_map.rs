use crate::model::NodeId;
use rayon::iter::{IntoParallelRefMutIterator, ParallelIterator};

const DEFAULT_NUMBER_OF_DIFFERENT_STATES_EXPECTED: usize = 64;

#[derive(Debug)]
pub struct StateMap(Vec<u8>);

impl StateMap {
    pub fn new() -> Self {
        Self(vec![0; DEFAULT_NUMBER_OF_DIFFERENT_STATES_EXPECTED])
    }

    pub fn count_states(&mut self, states: impl Iterator<Item = NodeId>) {
        // Clear vec; all states have 0 count
        self.0.par_iter_mut().for_each(|entry| *entry = 0);

        states
            .into_iter()
            .map(NodeId::as_index)
            .for_each(|idx| self.0[idx] += 1);
    }

    #[inline]
    pub fn get_count(&self, state: NodeId) -> u8 {
        self.0[state.as_index()]
    }

    pub const fn default_size() -> usize {
        size_of::<Self>() * DEFAULT_NUMBER_OF_DIFFERENT_STATES_EXPECTED
    }
}

impl Default for StateMap {
    fn default() -> Self {
        Self::new()
    }
}
