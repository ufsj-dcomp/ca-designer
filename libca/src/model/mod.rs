use std::collections::HashMap;

use edge::Edge;
use node::Node;
use serde::{Deserialize, Serialize};

pub use node::NodeId;

use crate::state_map::StateMap;

mod edge;
mod node;

#[cfg(test)]
pub mod test_utils;

#[derive(Serialize, Deserialize)]
pub struct Model {
    pub(crate) nodes: HashMap<NodeId, Node>,
    pub(crate) edges: Vec<Edge>,
}

impl Model {
    pub fn next_state(&self, curr_state: NodeId, neighbors: &StateMap) -> NodeId {
        self.edges
            .iter()
            .find_map(|edge| edge.transition(curr_state, neighbors))
            .unwrap_or(curr_state)
    }
}

#[cfg(test)]
mod tests {
    use test_utils::game_of_life_rules;

    use super::*;

    #[test]
    fn model_serialization() {
        insta::with_settings!({sort_maps =>true}, {
            insta::assert_ron_snapshot!(&game_of_life_rules());
        });
    }
}
