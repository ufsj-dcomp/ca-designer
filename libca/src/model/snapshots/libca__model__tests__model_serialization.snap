---
source: src/model/mod.rs
expression: "&model"
---
Model(
  nodes: {
    NodeId(0): Node("Dead"),
    NodeId(1): Node("Alive"),
  },
  edges: [
    Edge(
      id: EdgeId(0),
      name: "Underpopulation",
      from_node: NodeId(1),
      to_node: NodeId(0),
      conditions: [
        Condition(
          left: PopulationCount(NodeId(1)),
          operand: Less,
          right: Absolute(2),
        ),
      ],
    ),
    Edge(
      id: EdgeId(1),
      name: "Overpopulation",
      from_node: NodeId(1),
      to_node: NodeId(0),
      conditions: [
        Condition(
          left: PopulationCount(NodeId(1)),
          operand: Greater,
          right: Absolute(3),
        ),
      ],
    ),
    Edge(
      id: EdgeId(2),
      name: "Reproduction",
      from_node: NodeId(0),
      to_node: NodeId(1),
      conditions: [
        Condition(
          left: PopulationCount(NodeId(1)),
          operand: Equal,
          right: Absolute(3),
        ),
      ],
    ),
  ],
)
